src/initialize.cc

```diff
@@ -97,9 +97,7 @@ LibraryInitializer::LibraryInitializer()
   install_pthread_atfork_handlers();
 }
 
-LibraryInitializer::~LibraryInitializer() {
-  close_open_libs();
-}
+LibraryInitializer::~LibraryInitializer() = default;
```

Comment:
Can we add an assertion which checks if all handles have been closed during shutdown of mxnet? That could allow to catch a possible leak.

Reply:
This was removed since there are cases where (for external ops for example) an object is registered in an MXNet data structure like the operator registry and during shutdown the object is attempted to be destructed but its pointing to an object in the loaded library. This ended up causing a segfault. Without closing open handles, we let loaded libraries live longer than libmxnet.so and allow it to shutdown cleanly. 

This is why we changed `MXLoadLib` to return the handle to the library and call `dlclose` on the handle in Python. 

However, this still isnt an issue since on process exit the library will be closed by the OS anyway when it cleans up.

Reply:
Right, but hence an assertion or some other debug log to make the user aware that there are still unclosed handles. The idea is to give a hint during shutdown. If somebody then sees hundreds of unclosed handles, that could be a strong indicator of something being wrong - I'm not talking about automatic closing of the handle, just a message

Reply:
We cant check if handles are closed while libmxnet.so is closing. It would have to be checked elsewhere so the loaded libraries can live during libmxnet.so shutdown. If we put in a check it will print out to the user every time since the expectation is that the handles are closed after libmxnet.so

We're not talking about hundreds of handles, we're talking about one or maybe two libraries loaded by the user explicitly.

Reply:
Okay, I was not aware that the goal was that these libraries have a longer lifecycle than libmxnet.so itself. Thanks for elaborating